elements/lisk-dpos/test/unit/vote_weight_snapshot.spec.ts

```diff
@@ -0,0 +1,870 @@
+				expect(voteWeights[0].round).toEqual(13);
+				expect(voteWeights[0].delegates).toHaveLength(103);
+				expect(
+					additionalDelegates.every((delegate: Account) =>
+						voteWeights[0].delegates.find(
+							(d: Account) => d.address === delegate.address,
+						),
+					),
+				).toBeTrue();
+			});
+		});
+
+		describe('when there are more than 2 delegates who have received votes more than the threshold (1000 * 10^8)', () => {
+			let additionalDelegates: Account[];
+			beforeEach(async () => {
+				delegates = getDelegateAccounts(101);
+				for (const delegate of delegates) {
+					delegate.totalVotesReceived = randomBigInt(3000, 5000);
+					delegate.votes.push({
+						delegateAddress: delegate.address,
+						amount: delegate.totalVotesReceived,
+					});
+				}
+				additionalDelegates = getDelegateAccounts(300);
+				for (const delegate of additionalDelegates) {
+					delegate.totalVotesReceived = randomBigInt(1000, 2999);
+					delegate.votes.push({
+						delegateAddress: delegate.address,
+						amount: delegate.totalVotesReceived,
+					});
+				}
+				block = {
+					id: 'random-block',
+					timestamp: 10100,
+					height: 1030,
+					generatorPublicKey: forgers[0].publicKey,
+					reward: BigInt(500000000),
+					totalFee: BigInt(100000000),
+					transactions: [],
+				} as Block;
+				chainStub.dataAccess.getDelegates.mockResolvedValue([
+					...delegates,
+					...additionalDelegates,
+				]);
+
+				// Setup for missed block calculation
+				const forgedBlocks = forgers
+					.map((forger, i) => ({
+						generatorPublicKey: forger.publicKey,
+						height: 928 + i,
+					}))
+					.slice(0, 102);
+				chainStub.dataAccess.getBlockHeadersByHeightBetween.mockResolvedValue(
+					forgedBlocks,
+				);
+
+				// TODO: Remove this forgers list mock after new DPoS implementation
+				const mockedForgersList = JSON.stringify([
+					{
+						round: 10,
+						delegates: [...forgers.map(d => d.publicKey).slice(0, 102)],
+					},
+				]);
+
+				stateStore = new StateStoreMock([forgers[0]], {
+					[CONSENSUS_STATE_FORGERS_LIST_KEY]: mockedForgersList,
+				});
+			});
+
+			it('should snapshot all the delegates who has more than the threshold', async () => {
+				// Act
+				await dpos.apply(block, stateStore);
+
+				// Assert
+				const voteWeightsStr = await stateStore.consensus.get(
+					CONSENSUS_STATE_VOTE_WEIGHTS_KEY,
+				);
+				const voteWeights = JSON.parse(voteWeightsStr as string);
+				expect(voteWeights).toHaveLength(1);
+				expect(voteWeights[0].round).toEqual(13);
+				expect(voteWeights[0].delegates).toHaveLength(
+					delegates.length + additionalDelegates.length,
+				);
+				expect(
+					additionalDelegates.every((delegate: Account) =>
+						voteWeights[0].delegates.find(
+							(d: Account) => d.address === delegate.address,
+						),
+					),
+				).toBeTrue();
+			});
+		});
+
+		describe('when there are non-self voting delegate within the top active delegates', () => {
```

Comment:
I'm not sure that this should be tested here? shouldn't that be unit tested when testing the `weight` function? What I mean, is: technically there can't be any non-self voting delegates in the active delegates...

Reply:
what do you mean by `weight` function?
Also, isn't it possible to have non-self vote within top 101?

Reply:
I mean, you have the function
delegate weight = max (10*self-votes, totalVotes)  (or 0 if punished).

This should be tested. Then the selection mechanism is somewhat independent.

Yes, my mistake, you can have a delegate with 0 weight in the top 101.

Reply:
that function is called during the snapshot phase, so that's why it's tested here

Reply:
I updated to top 101 delegates